File: javascript/src/ingestion/csv-pipeline.js
import { readdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

/**
 * Parse a single line of CSV, respecting double-quoted fields that may
 * contain commas or escaped quotes ("").
 *
 * @param {string} line - A single CSV row as a string.
 * @returns {string[]} Array of field values.
 */
function parseCsvLine(line) {
  const fields = [];
  let current = "";
  let inQuotes = false;
  let i = 0;

  while (i < line.length) {
    const ch = line[i];

    if (inQuotes) {
      if (ch === '"') {
        // Look ahead: doubled quote is an escaped quote inside a quoted field
        if (i + 1 < line.length && line[i + 1] === '"') {
          current += '"';
          i += 2;
          continue;
        }
        // Otherwise the quoted section ends
        inQuotes = false;
        i++;
        continue;
      }
      current += ch;
      i++;
    } else {
      if (ch === '"') {
        inQuotes = true;
        i++;
      } else if (ch === ",") {
        fields.push(current);
        current = "";
        i++;
      } else {
        current += ch;
        i++;
      }
    }
  }

  // Push the last field
  fields.push(current);
  return fields;
}

/**
 * Read a single CSV file and return its headers and rows.
 *
 * @param {string} filePath - Absolute or relative path to the CSV file.
 * @returns {Promise<{ headers: string[], rows: string[][] }>} Parsed CSV data.
 */
export async function readCsvFile(filePath) {
  const content = await readFile(filePath, "utf-8");
  const lines = content
    .split(/\r?\n/)
    .filter((line) => line.trim().length > 0);

  if (lines.length === 0) {
    return { headers: [], rows: [] };
  }

  const headers = parseCsvLine(lines[0]);
  const rows = lines.slice(1).map((line) => parseCsvLine(line));

  return { headers, rows };
}

/**
 * Validate that all required columns are present in the headers.
 *
 * @param {string[]} headers - The header row of a CSV file.
 * @param {string[]} requiredColumns - Column names that must be present.
 * @returns {string[]} Array of missing column names (empty if all present).
 */
export function validateColumns(headers, requiredColumns) {
  const headerSet = new Set(headers.map((h) => h.toLowerCase().trim()));
  return requiredColumns.filter(
    (col) => !headerSet.has(col.toLowerCase().trim()),
  );
}

/**
 * Standardize header names: lowercase, trim whitespace, replace spaces
 * with underscores.
 *
 * @param {string[]} headers - Original header names.
 * @returns {string[]} Standardized header names.
 */
export function standardizeHeaders(headers) {
  return headers.map((h) => h.trim().toLowerCase().replace(/\s+/g, "_"));
}

/**
 * Trim whitespace from every string field in the rows.
 *
 * @param {string[][]} rows - Array of row arrays.
 * @returns {string[][]} New array with all fields trimmed.
 */
export function trimFields(rows) {
  return rows.map((row) => row.map((field) => field.trim()));
}

/**
 * Remove exact duplicate rows.
 *
 * @param {string[][]} rows - Array of row arrays.
 * @returns {{ uniqueRows: string[][], removedCount: number }} Deduplicated rows and count of removed duplicates.
 */
export function deduplicate(rows) {
  const seen = new Set();
  const uniqueRows = [];

  for (const row of rows) {
    const key = row.join("\x00");
    if (!seen.has(key)) {
      seen.add(key);
      uniqueRows.push(row);
    }
  }

  return {
    uniqueRows,
    removedCount: rows.length - uniqueRows.length,
  };
}

/**
 * Serialize a row into a CSV line, quoting fields that contain commas,
 * double quotes, or newlines.
 *
 * @param {string[]} row - Field values.
 * @returns {string} A single CSV-formatted line.
 */
function rowToCsvLine(row) {
  return row
    .map((field) => {
      if (
        field.includes(",") ||
        field.includes('"') ||
        field.includes("\n") ||
        field.includes("\r")
      ) {
        return '"' + field.replace(/"/g, '""') + '"';
      }
      return field;
    })
    .join(",");
}

/**
 * Run the full CSV ingestion and cleaning pipeline.
 *
 * 1. Reads all .csv files from inputDir.
 * 2. Validates required columns (if given), rejecting non-conforming files.
 * 3. Standardizes headers and trims fields.
 * 4. Merges rows from all valid files.
 * 5. Deduplicates across the merged set.
 * 6. Writes the cleaned output CSV.
 *
 * @param {object} options
 * @param {string} options.inputDir - Directory containing CSV files.
 * @param {string} options.outputPath - Path for the output CSV file.
 * @param {string[]} [options.requiredColumns] - Optional list of required column names.
 * @returns {Promise<{
 *   filesProcessed: string[],
 *   filesRejected: string[],
 *   rowsRead: number,
 *   rowsWritten: number,
 *   duplicatesRemoved: number
 * }>} Pipeline result summary.
 */
export async function runPipeline({ inputDir, outputPath, requiredColumns }) {
  const result = {
    filesProcessed: [],
    filesRejected: [],
    rowsRead: 0,
    rowsWritten: 0,
    duplicatesRemoved: 0,
  };

  // 1. Discover CSV files
  let entries;
  try {
    entries = await readdir(inputDir);
  } catch (err) {
    console.warn(`Could not read input directory: ${err.message}`);
    return result;
  }

  const csvFiles = entries
    .filter((f) => f.toLowerCase().endsWith(".csv"))
    .sort();

  if (csvFiles.length === 0) {
    console.info("No CSV files found in input directory.");
    await writeFile(outputPath, "", "utf-8");
    return result;
  }

  let mergedHeaders = null;
  const allRows = [];

  for (const fileName of csvFiles) {
    const filePath = join(inputDir, fileName);

    // 2. Read CSV
    let parsed;
    try {
      parsed = await readCsvFile(filePath);
    } catch (err) {
      const reason = `${fileName}: read error — ${err.message}`;
      console.warn(reason);
      result.filesRejected.push(reason);
      continue;
    }

    if (parsed.headers.length === 0) {
      const reason = `${fileName}: empty or malformed file`;
      console.warn(reason);
      result.filesRejected.push(reason);
      continue;
    }

    // 3. Validate required columns (before standardizing, compare case-insensitively)
    if (requiredColumns && requiredColumns.length > 0) {
      const missing = validateColumns(parsed.headers, requiredColumns);
      if (missing.length > 0) {
        const reason = `${fileName}: missing required columns — ${missing.join(", ")}`;
        console.warn(reason);
        result.filesRejected.push(reason);
        continue;
      }
    }

    // 4. Standardize and trim
    const stdHeaders = standardizeHeaders(parsed.headers);
    const trimmedRows = trimFields(parsed.rows);

    result.rowsRead += trimmedRows.length;

    // 5. Merge — use the headers from the first valid file as canonical
    if (mergedHeaders === null) {
      mergedHeaders = stdHeaders;
    } else {
      // Reorder columns of this file to match the canonical header order
      // (skip files whose standardized headers don't match the canonical set)
      const canonicalSet = new Set(mergedHeaders);
      const thisSet = new Set(stdHeaders);
      const extraCols = stdHeaders.filter((h) => !canonicalSet.has(h));
      const missingCols = mergedHeaders.filter((h) => !thisSet.has(h));

      if (extraCols.length > 0 || missingCols.length > 0) {
        const reason = `${fileName}: header mismatch — extra: [${extraCols.join(", ")}], missing: [${missingCols.join(", ")}]`;
        console.warn(reason);
        result.filesRejected.push(reason);
        continue;
      }

      // Reorder rows to match canonical header order if columns are in a different order
      if (stdHeaders.join(",") !== mergedHeaders.join(",")) {
        const indexMap = mergedHeaders.map((h) => stdHeaders.indexOf(h));
        for (const row of trimmedRows) {
          allRows.push(indexMap.map((i) => row[i] ?? ""));
        }
        result.filesProcessed.push(fileName);
        continue;
      }
    }

    allRows.push(...trimmedRows);
    result.filesProcessed.push(fileName);
  }

  // 6. Deduplicate
  const { uniqueRows, removedCount } = deduplicate(allRows);
  result.duplicatesRemoved = removedCount;
  result.rowsWritten = uniqueRows.length;

  // 7. Write output
  if (mergedHeaders) {
    const lines = [
      rowToCsvLine(mergedHeaders),
      ...uniqueRows.map(rowToCsvLine),
    ];
    await writeFile(outputPath, lines.join("\n") + "\n", "utf-8");
  } else {
    await writeFile(outputPath, "", "utf-8");
  }

  console.info(
    `Pipeline complete: ${result.filesProcessed.length} file(s) processed, ` +
      `${result.filesRejected.length} rejected, ` +
      `${result.rowsRead} rows read, ${result.rowsWritten} written, ` +
      `${result.duplicatesRemoved} duplicates removed.`,
  );

  return result;
}

File: python/src/data_platform_lab/ingestion/csv_pipeline.py
"""CSV ingestion and cleaning pipeline.

Reads multiple CSV files from a directory, validates, standardises,
deduplicates, and writes a single cleaned output CSV.
"""

from __future__ import annotations

import csv
import logging
from dataclasses import dataclass, field
from pathlib import Path

logger = logging.getLogger(__name__)


@dataclass
class PipelineResult:
    """Summary produced by a single pipeline run."""

    files_processed: list[str] = field(default_factory=list)
    files_rejected: list[str] = field(default_factory=list)
    rows_read: int = 0
    rows_written: int = 0
    duplicates_removed: int = 0


def read_csv_file(path: Path) -> tuple[list[str], list[list[str]]]:
    """Read a single CSV file and return (headers, rows).

    Raises :class:`ValueError` if the file is empty (no header row).
    """
    with path.open(newline="", encoding="utf-8") as fh:
        reader = csv.reader(fh)
        try:
            headers = next(reader)
        except StopIteration:
            raise ValueError(f"CSV file is empty: {path}") from None
        rows = list(reader)
    return headers, rows


def validate_columns(headers: list[str], required: list[str]) -> list[str]:
    """Return a list of required columns that are missing from *headers*."""
    header_set = set(headers)
    return [col for col in required if col not in header_set]


def standardize_headers(headers: list[str]) -> list[str]:
    """Lowercase, strip whitespace, and replace spaces with underscores."""
    return [h.strip().lower().replace(" ", "_") for h in headers]


def trim_fields(rows: list[list[str]]) -> list[list[str]]:
    """Strip leading/trailing whitespace from every field."""
    return [[cell.strip() for cell in row] for row in rows]


def deduplicate(
    rows: list[list[str]],
) -> tuple[list[list[str]], int]:
    """Remove exact-duplicate rows.

    Returns (unique_rows, count_removed).
    """
    seen: set[tuple[str, ...]] = set()
    unique: list[list[str]] = []
    for row in rows:
        key = tuple(row)
        if key not in seen:
            seen.add(key)
            unique.append(row)
    removed = len(rows) - len(unique)
    return unique, removed


def _write_csv(path: Path, headers: list[str], rows: list[list[str]]) -> None:
    """Write *headers* and *rows* to a CSV at *path*."""
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", newline="", encoding="utf-8") as fh:
        writer = csv.writer(fh)
        writer.writerow(headers)
        writer.writerows(rows)


def run_pipeline(
    input_dir: Path,
    output_path: Path,
    required_columns: list[str] | None = None,
) -> PipelineResult:
    """Orchestrate the full CSV ingestion pipeline.

    1. Glob ``*.csv`` in *input_dir*.
    2. Read, validate, standardise headers, and trim each file.
    3. Merge rows, deduplicate, write the cleaned output, and return a summary.
    """
    result = PipelineResult()
    csv_files = sorted(input_dir.glob("*.csv"))

    if not csv_files:
        logger.warning("No CSV files found in %s", input_dir)
        _write_csv(output_path, [], [])
        return result

    merged_headers: list[str] | None = None
    all_rows: list[list[str]] = []

    for csv_file in csv_files:
        try:
            raw_headers, raw_rows = read_csv_file(csv_file)
        except Exception:
            reason = f"{csv_file.name}: unable to read file"
            logger.warning(reason)
            result.files_rejected.append(reason)
            continue

        std_headers = standardize_headers(raw_headers)

        # Validate required columns against the *standardised* headers.
        if required_columns is not None:
            missing = validate_columns(std_headers, required_columns)
            if missing:
                reason = (
                    f"{csv_file.name}: missing required columns "
                    f"{missing}"
                )
                logger.warning(reason)
                result.files_rejected.append(reason)
                continue

        # Establish the canonical header set from the first valid file.
        if merged_headers is None:
            merged_headers = std_headers

        # Only keep rows whose width matches the header width.
        trimmed = trim_fields(raw_rows)
        valid_rows = [
            r for r in trimmed if len(r) == len(std_headers)
        ]
        skipped = len(trimmed) - len(valid_rows)
        if skipped:
            logger.info(
                "%s: skipped %d malformed rows (column count mismatch)",
                csv_file.name,
                skipped,
            )

        result.rows_read += len(raw_rows)
        all_rows.extend(valid_rows)
        result.files_processed.append(csv_file.name)

    if merged_headers is None:
        merged_headers = []

    unique_rows, dups = deduplicate(all_rows)
    result.duplicates_removed = dups
    result.rows_written = len(unique_rows)

    _write_csv(output_path, merged_headers, unique_rows)
    return result
